catalog/app/containers/Bucket/PackageDialog/FilesInput.tsx

```diff
@@ -86,32 +109,107 @@ export type FilesAction = tagged.InstanceOf<typeof FilesAction>
+  changed?: ChangedDict
+  conflict?: string
+}
+
+type AnyFile = ExistingFile | AddedFile
+
+type AddedItems = Record<string, AddedFile>
+
+type ExistingItems = Record<string, ExistingFile>
+
 export interface FilesState {
-  added: Record<string, LocalFile | Model.S3File>
+  added: AddedItems
   deleted: Record<string, true>
-  existing: Record<string, Model.PackageEntry>
+  existing: ExistingItems
   // XXX: workaround used to re-trigger validation and dependent computations
   // required due to direct mutations of File objects
   counter?: number
 }
 
-const addMetaToFile = (
-  file: Model.PackageEntry | LocalFile | Model.S3File,
-  meta?: Model.EntryMeta,
-) => {
+function cloneDomFile<F extends AnyFile>(file: F, omitProperty: string): F {
+  const fileCopy = new window.File([file as File], (file as File).name, {
+    type: (file as File).type,
+  })
+  const properties = ['conflict', 'meta', 'changed', 'hash']
+  properties.forEach((propName) => {
+    const property = Object.getOwnPropertyDescriptor(file, propName)
+    if (property?.value !== undefined && omitProperty !== propName) {
+      Object.defineProperty(fileCopy, propName, property)
+    }
+  })
+  return fileCopy as F
+}
+
+function setKeyValue<T>(key: string, value: T, file: AnyFile): AnyFile {
   if (file instanceof window.File) {
-    const fileCopy = new window.File([file as File], (file as File).name, {
-      type: (file as File).type,
-    })
-    Object.defineProperty(fileCopy, 'meta', {
-      value: meta,
-    })
-    Object.defineProperty(fileCopy, 'hash', {
-      value: (file as FileWithHash).hash,
-    })
+    const fileCopy = cloneDomFile(file, key)
+    if (value !== undefined) {
+      Object.defineProperty(fileCopy, key, {
+        value,
+      })
+    }
     return fileCopy
   }
-  return R.assoc('meta', meta, file)
+  return R.assoc(key, value, file)
+}
+
+function addFile<T extends AnyFile, M extends Record<string, AnyFile>>(
+  path: string,
+  file: T,
+  mainItems: M,
+  itemsToCheck?: Record<string, AnyFile>,
+) {
+  const resolvedName = resolveName(
+    path,
+    itemsToCheck ? { ...mainItems, ...itemsToCheck } : mainItems,
+  )
+  return {
+    ...mainItems,
+    [resolvedName]:
+      resolvedName === path
+        ? setKeyValue<undefined>('conflict', undefined, file)
+        : setKeyValue<typeof path>('conflict', path, file),
+  }
+}
+
+function renameFile(p1: string, p2: string, a: AddedItems, e: ExistingItems): AddedItems
+function renameFile(
+  p1: string,
+  p2: string,
+  e: ExistingItems,
+  a: AddedItems,
+  reverted?: boolean,
+): ExistingItems
+function renameFile(
+  oldPath: string,
+  newPath: string,
+  mainItems: Record<string, AnyFile>,
+  itemsToCheck: Record<string, AnyFile>,
+  reverted = false,
+) {
+  const file = mainItems[oldPath]
+  if (!file) return mainItems
+  const itemsWithOldNameRemoved = R.dissoc(oldPath, mainItems)
+  const changedFile = setKeyValue<ChangedDict | undefined>(
```

Comment:
there's no need to pass the type arguments explicitly, TS can infer those by itself

Reply:
And what type use for argument in function definition?

```ts
function setKeyValue(key: string, value: any, file: AnyFile): AnyFile {
```
or
```ts
function setKeyValue(key: string, value: string|ChangedDict|Model.EntryMeta, file: AnyFile): AnyFile {
```
?

I'm not sure, I have preferences now over these three variants, just asking

Reply:
i don't mean changing function signature, i mean don't pass the type args explicitly, bc TS is smart enough to infer them in most cases, i.e. just use `setKeyValue(...)` instead of `setKeyValue<ChangedDict | undefined>(...)`